### DISCUSSION TITLE
Question about `ephemera` usage

### DISCUSSION THREAD

**Discussion by tgrecojs** [Agoric/agoric-sdk#6799] [Q&A] [answered]
Throughout the code for IST Vault's, I've noticed the use of `provideEphemera` throughout various files. If I understand correctly, we can use `provideEphemera` to set new key-value pairs within a `Map`. Once set, it can be used to look up values associated with a given. 

While I believe to have a decent grasp on the use of ephemeral objects, I'm confused about the use of `Object.assign` shown below. 

https://github.com/Agoric/agoric-sdk/blob/c680b92375748c2608ae22bb26ad9a09d38aca90/packages/inter-protocol/src/vaultFactory/vaultHolder.js#L45-L58

I can't find any references to the `ephemera` identifier defined on line 53, so I am opening up this issue to get some context into its usage and why it is necessary to achieve the desired functionality.

**Comment by turadg** [comment, marked as answer]
`provideEphemera` is the function returned by `makeEphemeraProvider` which holds the "ephemeras" in a closure https://github.com/Agoric/agoric-sdk/blob/d7652cb32721a9922d453aed3205e1cff5bc84f3/packages/inter-protocol/src/contractSupport.js#L150

Each ephemera object holds references that can be lost and the object is keyed by whatever was passed to `provideEphemera`. In this case it's `holderId`. (Though in master [at the moment it's `subscriber`](https://github.com/Agoric/agoric-sdk/blob/master/packages/inter-protocol/src/vaultFactory/vaultHolder.js#L70) which is 1:1 with the holder object).

The object on line 53 isn't strictly necessary. Those lines could be,
```js
Object.assign(provideEphemera(holderId), { subscriber, publisher }); 
```

The values written into the ephemera object are read without binding an `ephemera` const. E.g.https://github.com/Agoric/agoric-sdk/blob/c680b92375748c2608ae22bb26ad9a09d38aca90/packages/inter-protocol/src/vaultFactory/vaultHolder.js#L78

**Reply by tgrecojs** [reply]
Ok, thanks! 

Seeing this pattern you linked to [here]([at the moment it's subscriber](https://github.com/Agoric/agoric-sdk/blob/master/packages/inter-protocol/src/vaultFactory/vaultHolder.js#L70) helped clear that up for me. 🙂

**Reply by tgrecojs** [reply]
I'm noticing that in vaultDirector the [ephemera](https://github.com/Agoric/agoric-sdk/blob/master/packages/inter-protocol/src/vaultFactory/vaultDirector.js#L65) object is [assigned values](https://github.com/Agoric/agoric-sdk/blob/master/packages/inter-protocol/src/vaultFactory/vaultDirector.js#L198-L207) and then is accessed directly throughout the file. 

This is an instance where I do see the `ephemera` object being used without any use of `provideEphemera` so I'm wondering if there is a specific reason as to why this is the case? Thanks again for the help! @turadg

**Reply by turadg** [reply]
The reason vaultDirector can avoid the provider is it's a singleton, so there's no need to distinguish instances. It can keep the value in module scope because that module is guaranteed to load once per vat. It's still a bit of a hack and we're cleaning it up before Vaults Release.

**Reply by tgrecojs** [reply]
Awesome. Thanks for the info.
